Feature/recruitments application

Jiwoong0322 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #85 
- 작업 요약: 공고 참여 기능 구현

## 📄 상세 내용
- [ ] 한 명의 사용자가 동일한 공고에 두 번 이상 지원할 수 없도록 제한합니다.
- [ ] 사용자가 POST /api/v1/recruitments/{uuid}/application 엔드포인트로 지원서 데이터를 전송하면, 새로운 Application 객체를 새엇ㅇ, 데이터베이스에 저장합니다.
- [ ] 지원서의 '스터디 경험 유무' 필드가 True일 경우에만 '구체적인 스터디 경험' 필드 작성을 강제합니다.
- [ ] 유효하지 않은 공고 uuid로 요청이 들어올 경우, 작업을 중단하고 에러를 반환합니다.
- [ ] 테스트 코드 작성 및 안정성 향상 : 테스트 실행 시 User 및 Recruitments 모델의 필수 필드 누락으로 발생하는 IntegrityError와, 시간대 정보 누락으로 발생하는 RuntimeWarning을 해결했습니다.
- [ ] isort, black을 실행하고 mypy, coverage run/report모두 통과했습니다.

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Jiwoong0322 committed:
✨ feat: 공고 참여 기능 추가

REQ-APLY-001 스터디 지원서 생성 API 기능 구현, 관련 테스트 코드 작성
주요 기능
 - 지원서 생성 엔드포인트 추가 (POST api/v1/recruitments/{uuid}/applications)
 - 중복 지원 방지 및 조건부 유효성 검사 등 비즈니스 로직 적용

아키텍처
 - Manager 패턴을 적용, 모델 관련 비즈니스 로직을 캡슐화하여 View의 역할을 명확히 함
 - API 관점에 따라 author_urls.py로 URL 파일을 분리하여 관리

 테스트 및 안정성
  - API 동작을 검증하는 전체 테스트 코드 작성
  - 테스트 실행 시 발생하던 IntegrityError, RuntimeWarning 오류 수정
  - Mypy 정적 분석을 통과하도록 전체 코드에 타입 힌트 적용

관련 이슈: #85

# Conflicts:
#	config/urls.py

Jiwoong0322 committed:
♻️ refactor: applications urls 호출방식 변경

applications의 endpoint prefix로 인해 recruitments_urls.py에서 선 처리 후 applications/recruitments_uuid_urls.py를 호출하도록 변경

관련 이슈: #85

SimJunSik commented on apps/applications/application_views.py:
```diff
@@ -0,0 +1,75 @@
+        except IntegrityError as e:
+            return Response(
+                {"error_code": "DUPLICATE_APPLICATION", "message": str(e)}, status=status.HTTP_403_FORBIDDEN
+            )
```
다른 app 들을 봤을 때 service 레이어를 쓰고 있어서
여기 applications app 에서도, view 에서 직접적으로 ORM ( DB 호출 ) 을 하지 않게 service 레이어를 사용하도록 해보면 좋을 것 같네요

Jiwoong0322 committed:
♻️ refactor: applications앱에 서비스 계층 도입

- 기존 View에서 처리하던 비즈니스 로직을 분리하기 위해 서비스 계층 추가
- View는 HTTP 요청/응답 처리, Service는 비즈니스 로직, Manager는 DB 쿼리를 담당하도록 역할 구분

관련 이슈: #85

Meoyoug commented on apps/applications/managers/application_managers.py:
```diff
@@ -0,0 +1,48 @@
+        """
+        # exists()를 사용하여 DB에 공고 지원 이력 데이터 존재 여부만 확인한다.
+        # 실제 데이터를 가져오지 않으므로 get()이나 filter()보다 효율적이다.
+        if self.filter(user=user, recruitment=recruitment).exists():
```
db 모델 매니저의 경우 create함수는 create에 관련된 로직만 작성되도록 이 if 문의 조회는 별도의 함수로 빼주는 게 좋을 것 같아요

Meoyoug commented on apps/applications/services/application_services.py:
```diff
@@ -0,0 +1,27 @@
+    ) -> Application:
+
+        # Manager를 통해 실제 DB 로직을 수행하고, 서비스는 그 흐름을 제어한다.
+        application = Application.objects.create_application(user=user, recruitment=recruitment, **validated_data)
```
create_application 메서드에 구현된 중복 데이터 조회 -> 생성 로직을 매니저에서 함수분리 후에 서비스 레이어 내에 호출함으로써 구현함이 맞다고 봅니다

Jiwoong0322 committed:
♻️ refactor: 공고 지원 생성 로직의 역할 및 책임 분리

- 중복 지원을 확인하는 조회 로직을 has_applied 메소드로 분리
- create_application 메소드는 객체 생성 책임만 갖도록 단순화
- 서비스 계층에서 중복 여부를 먼저 검사하도록 수정

관련 이슈: #85

Meoyoug reviewed (approved):
고생하셨습니다~ 머지할게용! 
빠른수정좋아요~
